certora/helpers/MerkleTree.sol

```diff
@@ -0,0 +1,132 @@
+pragma solidity ^0.8.0;
+
+struct Leaf {
+    address addr;
+    address reward;
+    uint256 value;
+}
+
+struct InternalNode {
+    bytes32 id;
+    bytes32 left;
+    bytes32 right;
+}
+
+struct Node {
+    bytes32 left;
+    bytes32 right;
+    address addr;
+    address reward;
+    uint256 value;
+    // hash of [addr, reward, value] for leaves, and [left.hash, right.hash] for internal nodes.
+    bytes32 hashNode;
+}
+
+contract MerkleTree {
+    /* STORAGE */
+
+    // The tree has no root because every node (and the nodes underneath) form a Merkle tree.
+    mapping(bytes32 => Node) internal tree;
+
+    /* MAIN FUNCTIONS */
+
+    function newLeaf(Leaf memory leaf) public {
+        // The following identifier is used as the key to create a new leaf.
+        // This ensures that the same pair of address and reward does not appear twice in the tree.
+        bytes32 id = keccak256(abi.encode(leaf.addr, leaf.reward));
+        Node storage node = tree[id];
+        require(id != 0, "id is the zero bytes");
+        require(isEmpty(node), "leaf is not empty");
+        require(leaf.value != 0, "value is zero");
+
+        node.addr = leaf.addr;
+        node.reward = leaf.reward;
+        node.value = leaf.value;
+        node.hashNode = keccak256(bytes.concat(keccak256(abi.encode(leaf.addr, leaf.reward, leaf.value))));
+    }
+
+    function newInternalNode(InternalNode memory internalNode) public {
+        Node storage node = tree[internalNode.id];
+        Node storage leftNode = tree[internalNode.left];
+        Node storage rightNode = tree[internalNode.right];
+        require(internalNode.id != 0, "id is zero bytes");
+        require(isEmpty(node), "node is not empty");
+        require(!isEmpty(leftNode), "left is empty");
+        require(!isEmpty(rightNode), "right is empty");
+        require(leftNode.hashNode <= rightNode.hashNode, "children are not pair sorted");
+
+        node.left = internalNode.left;
+        node.right = internalNode.right;
+        node.hashNode = keccak256(abi.encode(leftNode.hashNode, rightNode.hashNode));
+    }
+
+    /* PURE AND VIEW FUNCTIONS */
+
+    function isEmpty(Node memory node) public pure returns (bool) {
+        return node.left == 0 && node.right == 0 && node.addr == address(0) && node.reward == address(0)
+            && node.value == 0 && node.hashNode == 0;
+    }
+
+    function isEmpty(bytes32 id) public view returns (bool) {
+        return isEmpty(tree[id]);
+    }
+
+    function getLeft(bytes32 id) public view returns (bytes32) {
+        return tree[id].left;
+    }
+
+    function getRight(bytes32 id) public view returns (bytes32) {
+        return tree[id].right;
+    }
+
+    function getValue(address addr, address reward) public view returns (uint256) {
+        bytes32 id = keccak256(abi.encode(addr, reward));
+        return tree[id].value;
+    }
+
+    function getHash(bytes32 id) public view returns (bytes32) {
+        return tree[id].hashNode;
+    }
```

Comment:
This one is necessary: we need to be able to compare the hash of the root of the merkle tree and the root that is stored on the rewards distributor. This is useful for:
- the checker, it makes sure that the root published onchain comes from a well-formed Merkle tree
- the verification of the rewards distributor, because it stands only for well-formed Merkle trees and so we need to be able to specify that